pkg/controllers/disruption/types.go

```diff
@@ -67,8 +67,8 @@ func NewCandidate(ctx context.Context, kubeClient client.Client, recorder events
 	if node.MarkedForDeletion() {
 		return nil, fmt.Errorf("state node is marked for deletion")
 	}
-	// skip candidates that aren't initialized
-	if !node.Initialized() {
+	// skip candidates that haven't been initialized for a period time
+	if !node.Disruptable(clk) {
```

Comment:
I like the idea of introducing Disruptable. Should it eventually be controlable beyond this static 30s value?

We have the ability to ignore nodes for disruption based in the DoNotDisrupt annotation but there is not temporal control there. 

https://github.com/kubernetes-sigs/karpenter/issues/920#issuecomment-1874946264 mentions an issue where this 3rd party autoscaler prevents handing the instances off to karpenter because it will disrupt too quickly.

I wonder if there is room in the disruption controls for something like "do not disrupt within first 5m of life" or a concept like this. 

Not saying its something we need to solve in this pr just food for thought

Reply:
> I wonder if there is room in the disruption controls for something like "do not disrupt within first 5m of life"

We have this as an annotation that can get propagated onto the Node from the NodeClaim today. Seems like this might be a really nice place to do https://github.com/kubernetes-sigs/karpenter/issues/752 and push the `karpenter.sh/do-not-disrupt` annotation that can either be configured at the Pod spec level or at the NodeClaim level (which could be passed down from NodePool `spec.template.metadata.annotations` so you could do something like):

```yaml
spec:
  template:
     metadata:
       annotations:
          karpenter.sh/do-not-disrupt: 10m
```

> Should it eventually be controlable beyond this static 30s value

I think the annotation handling above gives us a nice path to this; however, the Node `karpenter.sh/do-not-disrupt` min lifetime differs a bit from the InitializationBuffer -- both in implementation and also in what it should be checking against. The `karpenter.sh/do-not-disrupt` with a time should be comparing the current time to the creation timestamp of that "thing" (whether that be a Pod or a NodeClaim) and then checking if the time since its creation timestamp is within the grace period associated with the `karpenter.sh/do-not-disrupt` annotation. In the case of the initialization buffer, I think we always need this to be a static value based off of the initialization time. 

The original intention of the initialization operation was to know when Karpenter could start to deprovision nodes and nodes could be considered "healthy" by the system. One issue here is that after a node becomes "healthy" to schedule nodes, pods may not schedule to it immediately (scheduler needs some time to see the pods, new node, and bind the pods as well as our watch on our nodes and the pods needs to see the binding). If someone has their `consolidationPolicy` set to `WhenEmpty` and they have their `consolidateAfter` value set to `0s`, now we have a race. We'll try to terminate new nodes that we create (which are valid for pods) a lot since we see that that node that is initialized doesn't have any pods scheduled to it (but might 2-5s later). 

Our current solve for that was to use scheduler nomination to know that pods were going to land on nodes. This worked okay, but 1) It's more complicated than a flat amount of time and 2) We didn't extend that scheduler nomination when we were launching nodes for disruption. This means that any nodes that we were launching as replacements wouldn't know that they were having pods scheduled against them, which could be a problem for the reasons stated above around race conditions.

Given all this, it seems we'll always need a flat buffer on top of our initialization time to ensuring that we aren't racing against the kube-scheduler

Reply:
Well said!